File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: Optional[TreeNode]
        """
        def construct(preorder , st_p , end_p , inorder , st_i, end_i):
            if st_i>end_i or st_p > end_p:
                return None
           
            node = TreeNode(preorder[st_p])
            ind = hashMap[preorder[st_p]]
            left_length = ind - st_i
            right_length = end_i - ind

            node.left = construct(preorder,st_p + 1 , st_p + left_length, inorder , st_i , ind - 1)
            node.right = construct(preorder, st_p +left_length + 1 ,end_p, inorder , ind + 1 , end_i)
            return node

        hashMap = {}
        n = len(preorder) 
        for i,x in enumerate(inorder):
            hashMap[x] = i
        return construct(preorder,0,n-1,inorder,0,n-1)
        
        
            
        

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer,Integer> map = new HashMap<>();

    public TreeNode construct(int st_p , int end_p , int st_i , int end_i ,int[] preorder, int[] inorder){
        if(st_p > end_p || st_i > end_i)
            return null;
        
        TreeNode node = new TreeNode(preorder[st_p]);
        int ind = map.get(node.val);
        int leftLen = ind - st_i;
        int rightLen = end_i - ind;
        node.left = construct(st_p + 1 ,st_p + leftLen,st_i , ind - 1,preorder,inorder);
        node.right = construct(st_p + leftLen + 1 ,end_p ,ind + 1,end_i ,preorder,inorder);
        return node;
        
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = inorder.length;
        for(int i = 0 ; i < n ; i++)
            map.put(inorder[i] , i);
        return construct(0,n - 1,0,n - 1,preorder,inorder);
        
        
    }
}